### ISSUE TITLE
TLS1.3: Stateless retry and CCS

### ISSUE THREAD

**Issue by tatsuhiro-t** [openssl/openssl#5235]
I'm implementing QUIC, and specifically its stateless retry.
The idea is that server which wants to perform client address validation sends HRR with some stuff in cookie which can restore some state so that server does not have to store any client related information (stateless).  See #4283 

#4283 added SSL_stateless(), but my server handles 0-RTT, so the first call must be SSL_read_early_data().  After sending HRR, a server close QUIC connection.  Client then sends CSS and CH to server.  Server which is stateless, sees this as a first message, and calls SSL_read_early_data().  openssl checks message type, and generates error because CCS is not appropriate for the first message.  But tls1.3 draft 23 explicitly allows this:

https://tools.ietf.org/html/draft-ietf-tls-tls13-23#section-4.2.2

>   When a server is operating statelessly it may receive an unprotected
   record of type change_cipher_spec between the first and second
   ClientHello (see Section 5).  Since the server is not storing any
   state this will appear as if it were the first message to be
   received.  Servers operating statelessly MUST ignore these records.

If I understand correctly, openssl should ignore CCS in this case.

However, if I were used SSL_stateless(), it might work (but I haven't tried it, because the first call should be SSL_read_early_data(), and we have no way to know when we call SSL_stateless because server is stateless).

**Comment by mattcaswell** [comment]
Hmmm...this should work (I wrote that wording in the spec!). I'll check.

On a side note: Why are you using middlebox compat mode in QUIC? It is on by default but doesn't seem relevant to QUIC. You should disable it by clearing the SSL_OP_ENABLE_MIDDLEBOX_COMPAT option.

On a side side note (!): Doh! I just realised I forgot to document that option. :-(

**Comment by tatsuhiro-t** [comment]
Thank you for quick reply.  I'll try that option.

**Comment by tatsuhiro-t** [comment]
Clearing SSL_OP_ENABLE_MIDDLEBOX_COMPAT makes openssl client not send CCS.  Thank you.

This should be an another issue, but if I use SSL_read_early_data() as a first call in server side, app_gen_cookie_cb set by SSL_CTX_set_cookie_generate_cb() is not called.  At the moment, client complains about ServerHello sent by server because cryptographic context doesn't match.
I probably think it is only called via SSL_stateless(), and I'm wondering how to correctly use SSL_stateless() for 0-RTT capable server.
Ideally, it would be nice to SSL_read_early_data() or SSL_accept() internally handle stateless retry case as well, and calls cookie generation, verification, and restore the saved context.

**Comment by mattcaswell** [comment]
I've been thinking about this and I'm not entirely sure I understand what you are trying to do. Either your server is operating statelessly, or it accepts early data. You cannot do both - the spec doesn't allow it. If you are operating statelessly then when a new connection comes in there will be no cookie in the ClientHello so the server will issue an HRR. The client then responds with a new ClientHello containing the cookie.

It is illegal to send early data after a ClientHello if there has been an HRR. It is also illegal to send a cookie in the first ClientHello:

    Clients MUST NOT use cookies in their initial ClientHello in subsequent connections.

So there is no circumstance in which early data and a stateless server will end up with early data being accepted.

**Comment by tatsuhiro-t** [comment]
Yes, we cannot send early data after second flight CH.

>  Either your server is operating statelessly, or it accepts early data. You cannot do both 

If I understand this correctly, it means that operator has to deploy 2 instances of server: 1 for stateless, and another for early data (and no cookie).  But in practice, https server listens on tcp/443, and cannot deploy both 2 instances at once.

**Comment by kaduk** [comment]
It is probably easier to think about it as the operator must decide whether any given endpoint should be stateless or should allow early data, as the two are mutually incompatible.

**Comment by kaduk** [comment]
I guess I did not read the initial post closely enough and only looked at the follow-up.  QUIC 0-RTT does not use TLS 1.3 early data directly but instead uses the TLS exporter to get keys to use for QUIC replayable protected packets.
With QUIC, if you're in address-validating mode, the TLS handshake will either be a resumption or go through the cookie-validating flow (and if you're not in address-validating mode, you don't need the cookie at all).  So when you get a client initial, the handshake message goes to SSL_stateless and either resumption succeeds or you get a HelloRetryRequest+cookie.
I don't remember what the behavior of SSL_stateless() is for a ClientHello that should be resumable, but I suspect it does not succeed the resumption and instead always generates a cookie.  @mattcaswell is that correct?  If so, we should probably change SSL_stateless to succeed when resuming.

**Comment by mattcaswell** [comment]
> I guess I did not read the initial post closely enough and only looked at the follow-up. QUIC 0-RTT does not use TLS 1.3 early data directly but instead uses the TLS exporter to get keys to use for QUIC replayable protected packets.

Except we do not currently allow the export of the early data secret. I think IIRC @tatsuhiro-t mentioned somewhere in another PR or issue the existence of patch to enable this.

> I don't remember what the behavior of SSL_stateless() is for a ClientHello that should be resumable, but I suspect it does not succeed the resumption and instead always generates a cookie. @mattcaswell is that correct? If so, we should probably change SSL_stateless to succeed when resuming.

Well I was about to write that you are correct, SSL_stateless() always requires a cookie. But I just tested it, and it apparently succeeds without a cookie in the resumption case. I'm not convinced that is correct behaviour. Surely the whole point of the cookie is to prove that the client is capable of receiving messages at the claimed source IP. The point is to prevent amplification attacks where an attacker can send relatively small ClientHello messages and have the server flood the spoofed source IP with large data packets.

In TCP this isn't an issue because the connection setup proves the client can receive anyway. In UDP this is more of a problem. A resumption does not prevent this kind of amplification attack - it could be a replay from an attacker.

**Comment by kaduk** [comment]
The idea is that you can put an address verification token in the session ticket as well as in the cookie, and the limited ticket lifetime provides some freshness guarantee to limit the spoofing damage.

Per https://tools.ietf.org/html/draft-ietf-quic-tls-09#section-7.2:

```
7.2.  NewSessionTicket Address Validation

   The ticket in the TLS NewSessionTicket message allows a server to
   provide a client with a similar sort of token.  When a client resumes
   a TLS connection - whether or not 0-RTT is attempted - it includes
   the ticket in the handshake message.  As with the HelloRetryRequest
   cookie, the server includes the address validation token in the
   ticket.  TLS provides the token it extracts from the session ticket
   to the transport when it asks whether source address validation is
   needed.
```

**Comment by mattcaswell** [comment]
> The idea is that you can put an address verification token in the session ticket as well as in the cookie, and the limited ticket lifetime provides some freshness guarantee to limit the spoofing damage.

Hmmm....but I don't think there is a mechanism for doing that at the moment? So we shouldn't be resuming based on a ticket without it. It sounds like QUIC will need this capability though.

**Comment by kaduk** [comment]
>  I don't think there is a mechanism for doing that at the moment? So we shouldn't be resuming based on a ticket without it. It sounds like QUIC will need this capability though.

I also don't think there's currently a mechanism in our tree for that.  But "shouldn't" is perhaps debatable; our current tickets are fine for most things we're doing, just not for the additional demands QUIC will place on them.  So, we should add it, but it doesn't have to be a fire drill.

**Comment by mattcaswell** [comment]
>  But "shouldn't" is perhaps debatable

I meant shouldn't in the context of an SSL_stateless() call. If we're doing SSL_stateless() we should not resume from a psk in the absence of this feature (because the whole point of SSL_stateless() is to validate the source address, and without this feature we can't do that for a PSK unless we send an HRR first)

**Comment by kaduk** [comment]
> > But "shouldn't" is perhaps debatable

> I meant shouldn't in the context of an SSL_stateless() call.

That, I can wholeheartedly agree with!

**Comment by tatsuhiro-t** [comment]
I chatted with other TLS stack developers about this.
First let's put early data aside because it would make things complicated.

NSS and picotls have API to generate cookie when generating HRR which is just generated to request different key share, and does not perform address validation.  And any server can pick up cookie from this HRR, and continue handshake statelessly .
So I think that it would be nice that SSL_accept would do similar things.  That is, SSL_accept should allow application to decide whether cookie is added, or not, and also SSL_accept can recover state from incoming cookie just like SSL_stateless.

quicwg is now discussing whether all HRR should  bear cookie, or not regardless of address validation:
https://github.com/quicwg/base-drafts/issues/1094

**Comment by mattcaswell** [comment]
> So I think that it would be nice that SSL_accept would do similar things. That is, SSL_accept should allow application to decide whether cookie is added, or not, and also SSL_accept can recover state from incoming cookie just like SSL_stateless.

This doesn't make a whole lot of sense to me. Either stateless operation is important or not. If you're doing SSL_accept then we're going to be storing state. So why would we need to generate a cookie in that case?

**Comment by tatsuhiro-t** [comment]
I think that cookie is not necessary if connection goes with the same SSL object.
Meanwhile, the draft does not prohibit to generate cookie in this case.  And other TLS libraries have a flexibility to add cookie with/without explicit stateless mode.
Depending on the outcome of https://github.com/quicwg/base-drafts/issues/1094, QUIC might demand TLS library to generate cookie when server just requests different key share.  If server satisfies the key share provided in client first flight, we don't need stateless operation.  With current OpenSSL API, in order to use cookie, we have to use SSL_stateless, but it always incurs additional round trip even if server does not want different key share, or any address validation.

**Comment by mattcaswell** [comment]
If I understand you correctly, if the server decides to issue an HRR due to a bad key_share, then you want to be able operate statelessly and have a cookie put into the HRR. But if the key_share is good then you want to be able to continue without worrying about the cookie?

The problem is, if you don't call SSL_stateless() but then the server decides to issue an HRR anyway, then you will have to store that SSL object away somewhere until the client comes back again, i.e. you have to store the state - so you get none of the benefits of operating statelessly.

One possibility is to have some option to SSL_stateless() so that if we don't need to issue an HRR for a bad key_share then we proceed as if we had received a good cookie. Does that make sense?

**Comment by tatsuhiro-t** [comment]
> If I understand you correctly, if the server decides to issue an HRR
  due to a bad key_share, then you want to be able operate statelessly
  and have a cookie put into the HRR. But if the key_share is good
  then you want to be able to continue without worrying about the
  cookie?

Yes.

> The problem is, if you don't call SSL_stateless() but then the
  server decides to issue an HRR anyway, then you will have to store
  that SSL object away somewhere until the client comes back again,
  i.e. you have to store the state - so you get none of the benefits
  of operating statelessly.

In order to make server stateless, handshate state is put into cookie.

> One possibility is to have some option to SSL_stateless() so that if
  we don't need to issue an HRR for a bad key_share then we proceed as
  if we had received a good cookie. Does that make sense?

It is better, but I think that current SSL_stateless is fine.
Operator who wants to force retry can use it.  Ideally, it would be
awesome if OpenSSL provides the knob to the application that cookie is
always sent in HRR with SSL_accept.  And it can be extended to
SSL_read_early (see below).

The reason why I need this in SSL_accept instead of SSL_stateless is
that it can handle the resumption.  If SSL_accept can issue stateless
cookie, we can support following use cases:

1. When client hello is good to go, and there is no need to do address
   validation, then we can proceed handshake normally without retry.

2. In case of bad key_share, sends HRR with stateless cookie in order
   to make retry stateless if application wants that.

3. SSL_accept can handle resumtpion.

As for API, picotols offers the following orthogonal knobs to the
application:

https://github.com/h2o/picotls/pull/103

1. whether or not retry is forced.
2. whether or not retry is stateless.

As for SSL_read_early, I'm not suggesting that I want to do 0-RTT with
stateless cookie.  When I write 0-RTT capable server, I have to use
SSL_read_early instead of SSL_accept.  Assuming that SSL_read_early
and and SSL_accept share code to handle HRR, it would be nice to the
above functions available in SSL_read_early too.

**Comment by mattcaswell** [comment]
> > The idea is that you can put an address verification token in the session ticket as well as in the cookie, and the limited ticket lifetime provides some freshness guarantee to limit the spoofing damage.
> 
> Hmmm....but I don't think there is a mechanism for doing that at the moment? So we shouldn't be resuming based on a ticket without it. It sounds like QUIC will need this capability though.

There now is a mechanism for storing arbitrary app data inside a ticket, so that particular issue has gone away.

I'm unsure about the way ahead with this issue. The QUIC requirements in this area seem very confusing and its not clear to me what needs to be done.

**Comment by tatsuhiro-t** [comment]
I'll try https://www.openssl.org/docs/manmaster/man3/SSL_CTX_generate_session_ticket_fn.html  Thank you.

TLS interface that QUIC requires is summarized in https://tools.ietf.org/html/draft-ietf-quic-tls-11#section-4.2
The details are rather scattered in transport and tls draft.
I think that the feedback from TLS stack implementer is extremely valuable to the working group, especially for its feasibility.

**Comment by richsalz** [comment]
So maybe we should close out all QUIC issues, and remove code added just for QUIC, since the design team is planning on making big changes.

That's a pity, that QUIC support won't be in 1.1.1, a long-term support release, but I think it's the right thing to do.

**Comment by kaduk** [comment]
> remove code added just for QUIC

How much was there, really?  The session ticket appdata change has value independent of our intention for QUIC to use it to include an address validation token, and the only other recent commit logs mentioning QUIC seem to be for logging exporter keys.

**Comment by tatsuhiro-t** [comment]
> That's a pity, that QUIC support won't be in 1.1.1, a long-term support release, but I think it's the right thing to do.

It is, but I respect your decision.

I don't think the revised QUIC handshake currently proposed is merged as is, but there is no strong push back against the fundamental changes which require TLS stack to expose handshake key and read/write raw TLS messages without TLS record at the moment (this is probably because most of the participants, organization or companies in QUIC development has their own TLS stack).  Still QUIC interop will happen with these changes once it is merged.  Do you think that it is possible to implement them in OpenSSL?  There are at least 4 QUIC implementations which use OpenSSL and take place in interop BTW.

**Comment by mattcaswell** [comment]
>  Do you think that it is possible to implement them in OpenSSL?

It should be possible, but I haven't looked at the requirements in detail. It won't happen in the 1.1.1 timeframe though.

**Comment by richsalz** [comment]
It will absolutely happen that OpenSSL will support QUIC.  It just won't happen in the next release. It will be in master for a while.

**Comment by richsalz** [comment]
Is there anything else that needs this capability?

**Comment by mattcaswell** [comment]
I'm wondering if this whole issue has "gone away" with the QUIC redesign - at least for the purposes of 1.1.1 anyway. Should we close this?

**Comment by tatsuhiro-t** [comment]
I think this can be closed because QUIC Stateless retry does not use HRR anymore in the latest draft.

**Comment by richsalz** [comment]
Closing this, since QUIC design changed and this was done to support QUIC.
